sdk/trace/span.go

```diff
@@ -205,11 +211,40 @@ func (s *recordingSpan) SetStatus(code codes.Code, description string) {
 // will be overwritten with the value contained in attributes.
 //
 // If this span is not being recorded than this method does nothing.
+//
+// If adding attributes to the span would exceed the maximum amount of
+// attributes the span is configured to have, the last added attributes will
+// be dropped.
 func (s *recordingSpan) SetAttributes(attributes ...attribute.KeyValue) {
 	if !s.IsRecording() {
 		return
 	}
-	s.copyToCappedAttributes(attributes...)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, a := range attributes {
+		// Ensure attributes conform to the specification:
+		// https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/common/common.md#attributes
+		if !a.Valid() {
+			continue
+		}
+
+		if _, exists := s.attributes[a.Key]; exists {
+			// Setting an attribute with the same key as an existing attribute
+			// SHOULD overwrite the existing attribute's value.
+			s.attributes[a.Key] = a.Value
+		} else {
+			if len(s.attributes) >= s.maxAttributes {
+				// For each unique attribute key, addition of which would
+				// result in exceeding the limit, the SDK MUST discard that
+				// key/value pair.
+				s.droppedAttributes++
```

Comment:
That's a good question. Looking at the specification it is unspecified how to count dropped attributes. The only thing I can find is how they are reported, not what constitutes a dropped attribute.

I was also wondering if we should be counting attributes dropped because they are invalid.

Given the unspecified state of dropped attributes I think we are free to make the decision ourselves how we want to count them. I think we should probably change the current strategy to be all attributes dropped (every invalid or duplicate) or only valid unique ones. I think the former strategy is going to better server users. Dropping attributes and not reporting them will result in lost information and may confuse users to think there are bugs in opentelemetry-go or their code. Reporting these attributes as dropped will still obscure the reason they are dropped, but it will at least report something about the action.

Reply:
I had the same question, but I also realized we would have to capture the attribute keys somewhere to be able to check this uniqueness which would in most cases negate the reason we are dropping.

Reply:
I've updated to count all attributes dropped (invalid or duplicate).